### PR TITLE
PID1 and PID2 TPA Support Added

### PR DESCRIPTION
**Pull request by borisbstyle**
Pid controller 1  and 2 became really popular and there are many requests for TPA implementation.

As the regular coupled principles can't be applied to this PID controller there is only PID reduction being done without coupling.
Check the code and give me some feedback of what you think.

### CODE REVIEW

**Comment by borisbstyle** [comment]
And this has just been successfully flight tested.

Will play more with excessive values

**Comment by borisbstyle** [comment]
Ok guys 100% verified with flight testing that it works like it should.

Try to overtune your rig with PID1 excessively to oscillations and add a lot of tpa with low tpa_breakpoint to verify it works best.
I have to say best TPA implementation I flew so far. Normally when using TPA you get change of rotation rates as well what makes the quad unpredictable. In this way only the PID gains are being changed, but flight behavious stays the same.

I will look for option to decouple the rates from PIDs also on another PID controllers. Especially the harakiri is interresting due to maincuthz implementation for D term.

@thenickdude What do you think? Any changes suggested? Should I for example remove TPA from yaw axis or keep it as it is?

**Comment by borisbstyle** [comment]
OK wait with commiting this as I will rework the code to be able to use it for PID2 as well.

**Comment by thenickdude** [comment]
Yeah that looks good to me, a nice straightforward change. Though I would go for a more explicit variable name if possible (longer identifiers are okay!).

**Comment by borisbstyle** [comment]
Check the latest commit

File: src/main/flight/pid.c [hunk 1]
```diff
@@ -182,9 +182,9 @@ static void pidLuxFloat(pidProfile_t *pidProfile, controlRateConfig_t *controlRa
         RateError = AngleRate - gyroRate;
 
         // -----calculate P component
-        PTerm = RateError * pidProfile->P_f[axis];
+        PTerm = RateError * pidProfile->P_f[axis] * PIDscaler[axis] / 100;
         // -----calculate I component
-        errorGyroIf[axis] = constrainf(errorGyroIf[axis] + RateError * dT * pidProfile->I_f[axis] * 10, -250.0f, 250.0f);
+        errorGyroIf[axis] = constrainf(errorGyroIf[axis] + RateError * dT * pidProfile->I_f[axis] * PIDscaler[axis] / 1000, -250.0f, 250.0f);
```
Comment (thenickdude) [inline src/main/flight/pid.c]: That should be / 10 rather than / 1000.

File: src/main/flight/pid.c [hunk 2]
```diff
@@ -55,7 +55,7 @@ int16_t axisPID[3];
 int32_t axisPID_P[3], axisPID_I[3], axisPID_D[3];
 #endif
 
-uint8_t dynP8[3], dynI8[3], dynD8[3];
+uint8_t dynP8[3], dynI8[3], dynD8[3], PIDscaler[3];
```
Comment (thenickdude) [inline src/main/flight/pid.c]: I would add a comment here where the variable is declared that it is a scale factor for pitch and roll PIDs which is derived from the throttle, and 100 = 1.0x scale.

**Comment by borisbstyle** [comment]
Look very good.....the formula is simplified
 / 100 * 10 
to
/ 1000

Yes I will add the comment. That is a good idea

**Comment by borisbstyle** [comment]
I commented that as well for future coders

File: src/main/flight/pid.c [hunk 3]
```diff
@@ -182,9 +183,9 @@ static void pidLuxFloat(pidProfile_t *pidProfile, controlRateConfig_t *controlRa
         RateError = AngleRate - gyroRate;
 
         // -----calculate P component
-        PTerm = RateError * pidProfile->P_f[axis];
-        // -----calculate I component
-        errorGyroIf[axis] = constrainf(errorGyroIf[axis] + RateError * dT * pidProfile->I_f[axis] * 10, -250.0f, 250.0f);
+        PTerm = RateError * pidProfile->P_f[axis] * PIDscaler[axis] / 100;
+        // -----calculate I component. Note that PIDscaler is divided by 1000, because it is simplified formule from the previous multiply by 10
+        errorGyroIf[axis] = constrainf(errorGyroIf[axis] + RateError * dT * pidProfile->I_f[axis] * PIDscaler[axis] / 1000, -250.0f, 250.0f);
```
Comment (ledvinap) [inline src/main/flight/pid.c]: `pidProfile->I_f[axis] * 10 * PIDscaler[axis] / 100` => `idProfile->I_f[axis] * PIDscaler[axis] / 10`

File: src/main/mw.c [hunk 4]
```diff
@@ -100,7 +100,7 @@ int16_t headFreeModeHold;
 int16_t telemTemperature1;      // gyro sensor temperature
 static uint32_t disarmAt;     // Time of automatic disarm when "Don't spin the motors when armed" is enabled and auto_disarm_delay is nonzero
 
-extern uint8_t dynP8[3], dynI8[3], dynD8[3];
+extern uint8_t dynP8[3], dynI8[3], dynD8[3], PIDscaler[3];
```
Comment (ledvinap) [inline src/main/mw.c]: This should be moved to some header file that is included both in mw.c and pid.c. As implemented now, if one declaration changes, compiler will issue no warning, leading to hard to find error.

**Comment by borisbstyle** [comment]
Thanks for noticing that! Awesome.

Considering the etra header file. I don't see it as a part of this PR. I think separate PR should be made as it has nothing to do with this feature. I am not familliar with the current file naming convenctions.

**Comment by borisbstyle** [comment]
@thenickdude @ledvinap 
One more question guys. My current implementation of TPA applies to all axis Roll, Pitch and yaw.
On the regular TPA implementation of PID 0,3,4 and 5 yaw is not included in the TPA.
What do you think of that? Should I exclude yaw from TPA or keep it as it is? I personally think yaw should be included as it is now, cause that also can cause oscillations on high throttle.

The fix is easy:
//PID scalers for PID1 and PID2
    if (axis == YAW) {
        PIDscaler[axis] = 100;
    }
    else {
        PIDscaler[axis] = prop2;
   }

**Comment by borisbstyle** [comment]
Ok flew many packs today with it and I am happy as it is. And I mean TPA applied to all 3 axis.

**Comment by borisbstyle** [comment]
Still decided to remove the yaw axis from TPA and changed the variable name to PIDweight

**Comment by atomiclama** [comment]
Nice one Boris. Initial testing looked good for me, I'll try and get some more time for a bit of proper testing.

**Comment by borisbstyle** [comment]
@ledvinap I added the new variable to the header file

@atomiclama 
It works really well. I gave it to a few testers around the world. So far nothing strange reported. I think I am finished with it and waiting for the merge from Dominic

**Comment by hydra** [comment]
@borisbstyle excellent, thanks for working on this.

**Pull request merged**